### PR TITLE
S3 backend

### PR DESCRIPTION
**Pull request by martindurant**
With bytes, text and avro support (not csv, no test files).
Read-only file system for now.

### CODE REVIEW

File: distributed/s3.py [hunk 1]
```diff
@@ -2,16 +2,22 @@
...
+import sys
 
 import boto3
-from botocore.handlers import disable_signing
+from botocore.exceptions import ClientError
 from tornado import gen
 
 from dask.imperative import Value
+from dask.base import tokenize
+from distributed.utils import read_block, seek_delimiter
```
Comment (mrocklin) [inline distributed/s3.py]: Use relative imports within source files.  See the `.utils` import you've added a few lines below.

File: distributed/s3.py [hunk 2]
```diff
@@ -26,148 +32,635 @@
 
 get_s3_lock = threading.Lock()
 
+def split_path(path):
+    path = path.lstrip('s3://')
```
Comment (mrocklin) [inline distributed/s3.py]: This isn't quite what you want:

```python
In [1]: 's3://sally'.lstrip('s3://')
Out[1]: 'ally'
```

I suggest using `startswith` and slicing.

Comment (mrocklin) [inline distributed/s3.py]: It'd also be great to see a doctest in this function showing what it does.

File: distributed/s3.py [hunk 3]
```diff
@@ -26,148 +32,635 @@
...
+        raise NotImplementedError('S3 implemented as read-only')
+
+    def rm(self, path, recursive=True):
+        raise NotImplementedError('S3 implemented as read-only')
+
+    def exists(self, path):
+        return bool(self.ls(path))
+
+    def copy(self, path1, path2):
+        raise NotImplementedError('S3 implemented as read-only')
```
Comment (mrocklin) [inline distributed/s3.py]: I would prefer to remove these from the interface for now.  Otherwise we're just setting users up for disappointment.

File: distributed/s3.py [hunk 4]
```diff
@@ -26,148 +32,635 @@
...
+    logger.debug("Read %d blocks of binary bytes from %s", len(names), fn)
     if lazy:
-        values = [Value(name, [{name: (read_content_from_keys, bucket_name,
-                                       key, anon)}])
-                  for name, key in zip(names, keys)]
+        executor._send_to_scheduler({'op': 'update-graph',
+                                     'tasks': {},
+                                     'dependencies': set(),
+                                     'keys': [],
+                                     'client': executor.id})
```
Comment (mrocklin) [inline distributed/s3.py]: This send_to_scheduler call is not necessary.  It was used in hdfs to communicate restrictions, which don't exist here.

File: distributed/s3.py [hunk 5]
```diff
@@ -26,148 +32,635 @@
...
+                s3pars=s3pars, delimiter=delimiter)
 
 
-def read_text(bucket_name, prefix='', path_delimiter='', encoding='utf-8',
-        errors='strict', lineterminator='\n', executor=None, anon=None,
-        collection=True, lazy=True, compression=None):
-    """
-    Read lines of text from S3
+def read_block_from_s3(filename, offset, length, s3pars={}, delimiter=None):
+    s3 = S3FileSystem(**s3pars)
```
Comment (mrocklin) [inline distributed/s3.py]: Just to verify, this is fast to do after the first time, yes?

File: distributed/s3.py [hunk 6]
```diff
@@ -26,148 +32,635 @@
...
+
+
+def avro_to_df(b, av):
+    """Parse avro binary data with header av into a pandas dataframe"""
+    import pandas as pd
+    return pd.DataFrame(data=avro_body(b, av))
+
+
+@gen.coroutine
+def _read_avro(path, executor=None, fs=None, lazy=True, **kwargs):
```
Comment (mrocklin) [inline distributed/s3.py]: Thoughts on removing all of this avro and csv stuff for now?

File: distributed/s3.py (hunk 5 above)
Comment (martindurant) [inline distributed/s3.py]: The S3FileSystem does cache connections and directory entries; but actually boto3 establishes new HTTP connections for everything it does.

File: distributed/tests/test_s3.py [hunk 7]
```diff
@@ -27,60 +27,154 @@
...
+    assert fn not in s3.ls('distributed-test/')
+    assert fn in s3.ls('distributed-test/nested/')
+    assert fn in s3.ls('distributed-test/nested')
+    assert fn in s3.ls('s3://distributed-test/nested/')
+    assert fn not in s3.glob('distributed-test/')
+    assert fn not in s3.glob('distributed-test/*')
+    assert fn in s3.glob('distributed-test/nested')
+    assert fn in s3.glob('distributed-test/nested/*')
+    assert fn in s3.glob('distributed-test/nested/file*')
+    assert fn in s3.glob('distributed-test/*/*')
```
Comment (mrocklin) [inline distributed/tests/test_s3.py]: These are great.  Thanks!

I recommend splitting some of these off into separate tests.  One for glob, one for ls, one for cat/head/tail, etc..  This helps us to quickly isolate what breaks when we make changes.

File: distributed/s3.py (hunk 6 above)
Comment (martindurant) [inline distributed/s3.py]: CSV I removed due to no tests (but appears to work)
Seems a pity, it works surprisingly easy. Could be kept elsewhere...

File: distributed/tests/test_s3.py [hunk 8]
```diff
@@ -27,60 +27,154 @@
...
+    fn = 'test/accounts.1.json'
+    data = files[fn]
+    with s3.open('/'.join([test_bucket_name, fn])) as f:
+        seek_delimiter(f, b'}', 0)
+        assert f.tell() == 0
+        f.seek(1)
+        seek_delimiter(f, b'}', 5)
+        assert f.tell() == data.index(b'}') + 1
+        seek_delimiter(f, b'\n', 5)
+        assert f.tell() == data.index(b'\n') + 1
```
Comment (mrocklin) [inline distributed/tests/test_s3.py]: Can we verify also that it will seek from current location.  So for example if we seek to `data.index(b'\n') + 1` and then call `seek_delimiter` again do we end up at the new location?

File: distributed/tests/test_s3.py [hunk 9]
```diff
@@ -27,60 +27,154 @@
...
+    yield e._start()
+    for bs in [5, 15, 45, 1500]:
+        futures = read_bytes(test_bucket_name+'/test/accounts*',
+                             lazy=False, blocksize=bs, delimiter=b'\n')
+        results = yield e._gather(futures)
+        res = [r for r in results if r]
+        assert all(r.endswith(b'\n') for r in res)
+        ourlines = b''.join(res).split(b'\n')
+        testlines = b"".join(files.values()).split(b'\n')
+        assert set(ourlines) == set(testlines)
```
Comment (mrocklin) [inline distributed/tests/test_s3.py]: I think if you use `b''.join(files[k] for k in sorted(files))` you'll be able to test order here as well

File: distributed/tests/test_s3.py [hunk 10]
```diff
@@ -27,60 +27,154 @@
...
+        ourlines = b"".join(results).split(b'\n')
+        testlines = b"".join(files.values()).split(b'\n')
+        assert set(ourlines) == set(testlines)
+    yield e._shutdown()
+
+
+@gen_cluster(timeout=60)
+def test_read_bytes_delimited(s, a, b):
+    e = Executor((s.ip, s.port), start=False)
+    yield e._start()
```
Comment (mrocklin) [inline distributed/tests/test_s3.py]: If you're interested, I've added an `executor=True` keyword argument to `gen_cluster` to automate the construction and shutdown of the exectuor:

```python
@gen_cluster(timeout=60, executor=True)
def test_read_bytes_delimited(e, s, a, b):
   ...
```

File: distributed/utils.py [hunk 11]
```diff
@@ -352,3 +352,115 @@ def convert(task):
...
+
+        offset = start
+        length = end - start
+
+        # if length and not eof:
+        #    length -= len(delimiter)
+
+    f.seek(offset)
+    bytes = f.read(length)
+    return bytes
```
Comment (mrocklin) [inline distributed/utils.py]: This function could use its test companion from the hdfs3 library

**Comment by mrocklin** [comment]
This is looking great to me so far

File: distributed/s3.py (hunk 6 above)
Comment (mrocklin) [inline distributed/s3.py]: Yeah, we should keep working on it in another branch and merge it when it's ready.

File: distributed/utils.py (hunk 11 above)
Comment (martindurant) [inline distributed/utils.py]: Sorry, what do you mean?

Comment (mrocklin) [inline distributed/utils.py]: We should copy over this test from hdfs3: https://github.com/dask/hdfs3/blob/master/hdfs3/tests/test_utils.py#L32-L52

If a future developer decides to change this function then something needs to yell at them.

File: distributed/hdfs.py [hunk 12]
```diff
@@ -78,7 +80,7 @@ def read_bytes(fn, executor=None, hdfs=None, lazy=True, delimiter=None,
         offsets = [max([o, 1]) for o in offsets]
     lengths = [d['length'] for d in blocks]
     workers = [[h.decode() for h in d['hosts']] for d in blocks]
-    names = ['read-binary-%s-%d-%d' % (fn, offset, length)
+    names = ['read-binary-hdfs3-%s-%s' % (fn, tokenize(offset, length, delimiter, not_zero))
```
Comment (mrocklin) [inline distributed/hdfs.py]: Can we add a regression test for this?  Perhaps something like 

```python
A = read_bytes(..., delimiter=b'\n')
B = read_bytes(..., delimiter=b'foo')
assert [a.key for a in A] != [b.key for b in B]
```

File: distributed/s3.py [hunk 13]
```diff
@@ -26,148 +32,526 @@
...
+    >>> split_path("s3://mybucket/path/to/file")
+    ("mybucket", "path/to/file")
+    """
+    if path.startswith('s3://'):
+        path = path[5:]
+    if '/' not in path:
+        return path, ""
+    else:
+        bits = path.split('/')
+        return bits[0], '/'.join(bits[1:])
```
Comment (mrocklin) [inline distributed/s3.py]: I think this can be `path.split('/', 1)`

```python
In [1]: '1,2,3,4,5'.split(',')
Out[1]: ['1', '2', '3', '4', '5']

In [2]: '1,2,3,4,5'.split(',', 1)
Out[2]: ['1', '2,3,4,5']
```

File: distributed/s3.py [hunk 14]
```diff
@@ -26,148 +32,526 @@
...
-        Name of S3 bucket like ``'my-bucket'``
-    prefix: string
-        Prefix of key name to match like ``'/data/2016/``
-    path_delimiter: string (optional)
-        Delimiter like ``'/'`` to define implicit S3 directory structure
+    fn: string
+        location in S3
     executor: Executor (optional)
         defaults to most recently created executor
+    s3: HDFileSystem (optional)
```
Comment (mrocklin) [inline distributed/s3.py]: typo

File: distributed/s3.py (hunk 13 above)
Comment (martindurant) [inline distributed/s3.py]: Ah, I undid that because `maxsplit` is not in py2

File: distributed/utils.py (hunk 11 above)
Comment (mrocklin) [inline distributed/utils.py]: ping

File: distributed/s3.py (hunk 13 above)
Comment (mrocklin) [inline distributed/s3.py]:
```python
In [1]: '1,2,3,4,5'.split(',', 1)
Out[1]: ['1', '2,3,4,5']

In [2]: import sys; sys.version
Out[2]: '2.6.9 |Continuum Analytics, Inc.| (unknown, Aug 21 2014, 18:28:52) \n[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]'
```

File: distributed/s3.py [hunk 15]
```diff
@@ -26,148 +32,526 @@
...
+            files = self._ls(path)
+        except ClientError:
+            files = []
+        if path:
+            pattern = re.compile(path + '/[^/]*.$')
+            files = [f for f in files if pattern.match(f['Key']) is not None]
+            if not files:
+                try:
+                    files = [self.info(path)]
+                except OSError:
```
Comment (mrocklin) [inline distributed/s3.py]: In Python 2 this raises an IOError

Comment (martindurant) [inline distributed/s3.py]: Yes, that's a bug, should be IOError

File: distributed/tests/test_s3.py [hunk 16]
```diff
@@ -27,60 +27,171 @@
...
+        seek_delimiter(f, b'}', 5)
+        assert f.tell() == data.index(b'}') + 1
+        seek_delimiter(f, b'\n', 5)
+        assert f.tell() == data.index(b'\n') + 1
+        f.seek(1, 1)
+        ind = data.index(b'\n') + data[data.index(b'\n')+1:].index(b'\n') + 1
+        seek_delimiter(f, b'\n', 5)
+        assert f.tell() == ind + 1
+
+
```
Comment (mrocklin) [inline distributed/tests/test_s3.py]: This test is slow.  We should decorate it with the `slow` decorator in `utils_test.py`.

Alternatively, it would be interesting to learn to use `moto`, which would really help with things like this.  If you want to dive into S3 more seriously it might be worth investing some time here.

File: distributed/s3.py (hunk 15 above)
Comment (mrocklin) [inline distributed/s3.py]: It's probably both.  This is a common point of difference between Python 2 and 3.

I recommend `except (OSError, IOError):`

File: distributed/tests/test_s3.py (hunk 16 above)
Comment (martindurant) [inline distributed/tests/test_s3.py]: I don't think I am interested in doing more with S3 than we currently have in scope. As you said, though, tests should be much faster within AWS.

Comment (mrocklin) [inline distributed/tests/test_s3.py]: Moto would also facilitate things like writing and would make our tests more self-describing, rather than relying on pre-seeded buckets.

I'm just thinking aloud here about moto though.  I think we should get this in and then think about using moto as we expand towards writing and more filesystem-y features.

File: distributed/utils.py (hunk 11 above)
Comment (martindurant) [inline distributed/utils.py]: Yes, so we want to keep that final delimiter - the test needs to be modified.

Comment (mrocklin) [inline distributed/utils.py]: OK, either way it should be explicit.  We need to include that test within this PR.

File: distributed/hdfs.py [hunk 17]
```diff
@@ -21,7 +22,7 @@
 
 
 def walk_glob(hdfs, path):
-    if '*' not in path and hdfs.info(path)['kind'] == 'directory':
+    if '*' not in path and hdfs.info(path).get('kind', '') == 'directory':
```
Comment (mrocklin) [inline distributed/hdfs.py]: What brought this up?

Comment (mrocklin) [inline distributed/hdfs.py]: To be pedantic here, in an ideal world no source change should really ever happen without a similar change in tests.  Is it ok if I revert this change?  If not why not?

File: distributed/s3.py [hunk 18]
```diff
@@ -26,148 +32,525 @@
...
+            if bucket == '':
+                # list of buckets
+                if self.anon:
+                    # cannot list buckets if not logged in
+                    return []
+                files = self.s3.list_buckets()['Buckets']
+                for f in files:
+                    f['Key'] = f['Name']
+                    f['Size'] = 0
+                    del f['Name']
```
Comment (mrocklin) [inline distributed/s3.py]: What is the reasoning behind this?

File: distributed/s3.py [hunk 19]
```diff
@@ -26,148 +32,525 @@
...
+                self.cache = self.s3.s3.get_object(Bucket=self.bucket, Key=self.key,
+                                                Range='bytes=%i-%i' % (start, self.end - 1)
+                                                )['Body'].read()
+            if start < self.start:
+                new = self.s3.s3.get_object(Bucket=self.bucket, Key=self.key,
+                                         Range='bytes=%i-%i' % (start, self.start - 1)
+                                         )['Body'].read()
+                self.start = start
+                self.cache = new + self.cache
+            if end > self.end:
```
Comment (mrocklin) [inline distributed/s3.py]: When does this ever occur?  I can't seem to trigger this code path.

File: distributed/hdfs.py (hunk 17 above)
Comment (martindurant) [inline distributed/hdfs.py]: My apologies, this should be reverted - it is a remnant of trying to get the hdfs read_bytes to work with s3fs directly

Comment (mrocklin) [inline distributed/hdfs.py]: It was reverted (or rather I just didn't bring along the change) in #162

File: distributed/s3.py (hunk 18 above)
Comment (martindurant) [inline distributed/s3.py]: To make a user's buckets look like top-level directories. That normalises (to my mind) a path like s3://mybucket/path/file ; you can do a ls('/') and see apparent directories.

File: distributed/s3.py (hunk 19 above)
Comment (martindurant) [inline distributed/s3.py]: You can request any block, including past the end of the file (like with BytesIO); that should return empty bytestrings and not error.

Comment (mrocklin) [inline distributed/s3.py]: I wasn't able to trigger this.  What is a code snippet that might trigger this code?

Comment (martindurant) [inline distributed/s3.py]: Here is the snippet - and it broke! `self.s3.s3.get_object.get(` should b `self.s3.s3.get_object(`

```python
s3 = distributed.s3.S3FileSystem()
f = s3.open('distributed-test/test/accounts.1.json', block_size=20)
while f.read(10):
    pass
f.seek(5000)
f.read(10)
```

Each time you go past the current (very small) cache, you need that code path; also seeking past the end of the file. `read_block` explicitly checks the file size instead.

Comment (mrocklin) [inline distributed/s3.py]: Great.  Added in https://github.com/mrocklin/distributed/commit/de7550b847db1e7b12f739962baafad5fd153afd

**Comment by mrocklin** [comment]
Stale, closing
